File: basic_dsa_three_hundred_eight.java
/**
 * Basic DSA Problems - Best Time to Buy and Sell Stock
 * Find max profit from one buy and one sell
 */
public class basic_dsa_three_hundred_eight {

    // Problem: Array prices where prices[i] is stock price on day i.
    // Return max profit from one buy and one sell. If no profit, return 0.

    /**
     * Track min price seen so far; at each day, profit = price - min.
     * Time O(n), Space O(1)
     */
    public static int maxProfit(int[] prices) {
        int min = Integer.MAX_VALUE, profit = 0;
        for (int p : prices) {
            min = Math.min(min, p);
            profit = Math.max(profit, p - min);
        }
        return profit;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Best Time to Buy and Sell Stock");
        System.out.println("=".repeat(60));

        System.out.println("[7,1,5,3,6,4] -> " + maxProfit(new int[]{7, 1, 5, 3, 6, 4}));
        System.out.println("[7,6,4,3,1] -> " + maxProfit(new int[]{7, 6, 4, 3, 1}));
        System.out.println("[2,4,1] -> " + maxProfit(new int[]{2, 4, 1}));
    }
}

File: basic_dsa_three_hundred_eight.py
"""
Basic DSA - Best Time to Buy and Sell Stock with Cooldown
You may complete as many transactions as you like, but after you sell stock,
you cannot buy stock on the next day (cooldown 1 day).
Return the maximum profit.
"""


def max_profit_with_cooldown(prices):
    """
    DP with 3 states:
    hold: max profit while holding a stock
    sold: max profit just sold today
    rest: max profit in cooldown/rest state
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not prices:
        return 0

    hold = -prices[0]
    sold = 0
    rest = 0

    for p in prices[1:]:
        prev_sold = sold
        sold = hold + p
        hold = max(hold, rest - p)
        rest = max(rest, prev_sold)

    return max(sold, rest)


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Best Time to Buy and Sell Stock with Cooldown")
    print("=" * 60)

    print(max_profit_with_cooldown([1, 2, 3, 0, 2]))  # 3
    print(max_profit_with_cooldown([1]))  # 0
